### ISSUE TITLE
conflicting behavior in slice vs. spread test

### ISSUE THREAD

**Issue by nazrhyn** [psiho/jsbench-me#11]
## The Test
**Setup**
```js
const ray = [];
for (let i = 0; i < 10000; i++) {
	ray[i] = i.toString();
}
```
**Case 1: Array#slice**
```js
ray.slice();
```
**Case 2: Array spread**
```js
void [...ray];
```

## The Issue
The result of this benchmark on **jsbench.me** differs from other sites using [Benchmark.js](https://benchmarkjs.com/), and from some local code I wrote to test it using Benchmark.js, manually. In all other tests, the **Array#slice** case wins, but on **jsbench.me**, it loses. Additionally, the `hz` value displayed on **jsbench.me** is different than in the other tests, where the values are in the same general ballpark.

I'm not sure what the difference is, but it seems possible that `setup` is not being registered properly? You have the code I wrote to test this locally in a gist, below, so you can compare it with what you've written.

![image](https://user-images.githubusercontent.com/1672405/34613980-e00c22a4-f1fd-11e7-8205-d5a8d21f80b8.png)

### jsperf.com
![image](https://user-images.githubusercontent.com/1672405/34613751-f2ac031c-f1fc-11e7-932c-827ca760a8ea.png)
### jsbench.github.io
![image](https://user-images.githubusercontent.com/1672405/34613799-1c937084-f1fd-11e7-87ec-bf846e26f798.png)
### jsben.ch
![image](https://user-images.githubusercontent.com/1672405/34613857-5a39ed32-f1fd-11e7-8697-30f7ebc7a3e9.png)
### Local Code
<sup>Source at this gist: [Array#slice vs. Array spread](https://gist.github.com/nazrhyn/c884b8a76e98fca738e54d5ca2f51830)</sup>
```
$ node benchmark.js 
Running benchmark...
Array#slice x 41,124 ops/sec ±8.20% (87 runs sampled)
Array spread x 2,248 ops/sec ±0.53% (94 runs sampled)

Array#slice is fastest!

Done.
```

**Comment by psiho** [comment]
This is a most interesting find. Thank you for reporting it in so much detail. At the moment I'm working to make JSBenchme work with MS browsers, which is the most important step to move from beta to pre-release version. I'll try to squeeze this into next release if I find the cause fast enough. Otherwise, it will be right at the top of the list after that.

**Comment by nazrhyn** [comment]
My pleasure :wink:. I quite enjoy writing descriptive issue reports.

Let me know if there's anything I can help you with; I'll be watching this for comments.

**Comment by psiho** [comment]
Just a quick update. It took me yesterday 2 hours investigating this and finally I decided to publish 0.4 versio to support MS browsers first.

Today, after 2 more hours of banging my head, I'm still lost. Behaviour is crazy. I rewrote your test script for browser and run it localy... same result, error is there. But copying most of it to my full app, replacing complete sandbox logic in jsbenchme with similar (almost static) file and thus disconnecting it with the app itself (except it runs in sandboxed iframe of the app) produces inconsistent result. Sometimes correct and sometimes incorrect. Restarting browser sometimes helps, restarting wepack server sometimes help, and sometimes nothing :)

For now, all I know is that ths is some weird behavior of Chrome in combination with my webpack/React setup. I tried the online version with FF and there is no issue there. So it must be related to Chrome.

I'll have to continue tomorrow.

**Comment by nazrhyn** [comment]
The full source code for https://jsbench.github.io is at https://github.com/jsbench/jsbench.github.io. You could compare their implementation to yours/

**Comment by psiho** [comment]
Uh, I think I got it. Issues was with Mobx library. Was stripping down the code to see when it stops breaking and almost went crazy because very little code was left to delete :) At the end, importing Mobx library was enough to break it! And it seems only for this weird test case.

I've upgraded Mobx and some other libraries and first test are ok now. I'll prepare a new publish probably tomorrow.

This was a tough one.

**Comment by nazrhyn** [comment]
:tophat: Hats off to you! Why the heck would a state-management library cause something like that to happen??

**Comment by psiho** [comment]
That's what made this problem so difficult! No idea! And really weird that only some (so far I'm aware of only one) tests behave like this!

Further... it's not only state-management library! Tests are run in completely isolated sandboxed iframe with only benchmark.js and other libraries just like in your test case, no Mobx anywhere inside iframe! Mobx is on parent document (React/ Mobx) which is completely isolated, only postMessage() communication is allowed.

Even further... I deleted all React/Mobx components from parent doc, all code, only includes at the top were left. So React/Mobx wasn't even running, only included, and blank HTML page with that same iframe was output.... same error! It''s only after removing `import { observer } from 'mobx'` completely did the trick. So, just importing/Webpacking Mobx caused this, without even runing it, and only in Chrome.

**Comment by psiho** [comment]
... anyway. Fix is published. v0.4.2-beta is out. I'm still keeping this in beta status till I get some cross-browser reports.

**Comment by nazrhyn** [comment]
Whew. I wonder if Mobx is doing some kind of polyfilling or monkey patching that it probably shouldn't be doing? That would be the only explanation for changes happening when only importing -- stuff it does when the module code is executed.

Have you done any Microsoft browser testing, yourself? I switched my workstation to Linux at work a while back, so I definitely couldn't without these nice VM images Microsoft provides for testing their browsers:   
https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

![image](https://user-images.githubusercontent.com/1672405/34783006-552545d6-f5f8-11e7-9f26-57fd38f5f2ff.png)

Figured I'd mention that in case you didn't know about it!
